clients/filesystem-hadoop/src/main/java/com/datastrato/gravitino/filesystem/hadoop/GravitinoVirtualFileSystem.java

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright 2024 Datastrato Pvt Ltd.
+ * This software is licensed under the Apache License version 2.
+ */
+package com.datastrato.gravitino.filesystem.hadoop;
+
+import com.datastrato.gravitino.Catalog;
+import com.datastrato.gravitino.NameIdentifier;
+import com.datastrato.gravitino.client.FilesetCatalog;
+import com.datastrato.gravitino.client.GravitinoClient;
+import com.datastrato.gravitino.client.GravitinoMetaLake;
+import com.datastrato.gravitino.file.Fileset;
+import com.datastrato.gravitino.shaded.com.google.common.base.Preconditions;
+import com.datastrato.gravitino.shaded.org.apache.commons.lang3.StringUtils;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Arrays;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FSDataOutputStream;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.InvalidPathException;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.permission.FsPermission;
+import org.apache.hadoop.util.Progressable;
+
+public class GravitinoVirtualFileSystem extends FileSystem {
+  private FileSystem proxyFileSystem;
+  private Path workingDirectory;
+  private URI uri;
+  private NameIdentifier filesetIdentifier;
+
+  @Override
+  public void initialize(URI name, Configuration conf) throws IOException {
+    if (name.toString().startsWith(GravitinoVirtualFileSystemConfiguration.GVFS_FILESET_PREFIX)) {
+      Preconditions.checkArgument(
+          StringUtils.isNotBlank(
+              conf.get(GravitinoVirtualFileSystemConfiguration.FS_GRAVITINO_SERVER_URI_KEY)),
+          "Gravitino server uri is not set in the configuration");
+      this.filesetIdentifier = normalizedIdentifier(name.getPath());
```

Comment:
This also increase the restriction of the FileSystem. We must initialize the FileSystem with fileset. But for other FileSystem, they can initialize with any path.